nilearn/datasets/neurovault.py

```diff
@@ -0,0 +1,2589 @@
+
+        If ``self.fetch_ns_`` is set and Neurosynth words are not on
+        disk, fetch them and add their location to image metadata.
+
+        """
+        if not self.write_ok_:
+            return image_info
+        image_info = self._add_words(image_info)
+        metadata_file_path = os.path.join(
+            os.path.dirname(image_info['absolute_path']),
+            'image_{0}_metadata.json'.format(image_info['id']))
+        _write_metadata(image_info, metadata_file_path)
+        return image_info
+
+    def start(self):
+        """Prepare for a download session.
+
+        If we don't have a sandbox directory for downloads, create
+        one.
+
+        """
+        if self.temp_dir_ is None:
+            self.temp_dir_ = _get_temp_dir(self.suggested_temp_dir_)
+
+    def finish(self):
+        """Cleanup after download session.
+
+        If ``self.start`` created a temporary directory for the
+        download session, remove it.
+
+        """
+        if self.temp_dir_ is None:
+            return
+        if self.temp_dir_ != self.suggested_temp_dir_:
+            shutil.rmtree(self.temp_dir_)
+            self.temp_dir_ = None
+
+
+def _json_from_file(file_name):
+    """Load a json file encoded with UTF-8."""
+    with open(file_name, 'rb') as dumped:
+        loaded = json.loads(dumped.read().decode('utf-8'))
+    return loaded
+
+
+def _json_add_collection_dir(file_name, force=True):
+    """Load a json file and add is parent dir to resulting dict."""
+    loaded = _json_from_file(file_name)
+    set_func = loaded.__setitem__ if force else loaded.setdefault
+    dir_path = os.path.dirname(file_name)
+    set_func('absolute_path', dir_path)
+    set_func('relative_path', os.path.basename(dir_path))
+    return loaded
+
+
+def _json_add_im_files_paths(file_name, force=True):
+    """Load a json file and add image, reduced rep and words paths."""
+    loaded = _json_from_file(file_name)
+    set_func = loaded.__setitem__ if force else loaded.setdefault
+    dir_path = os.path.dirname(file_name)
+    dir_relative_path = os.path.basename(dir_path)
+    image_file_name = 'image_{0}.nii.gz'.format(loaded['id'])
+    reduced_file_name = 'image_{0}_reduced_rep.npy'.format(loaded['id'])
+    words_file_name = 'neurosynth_words_for_image_{0}.json'.format(
+        loaded['id'])
+    set_func('relative_path', os.path.join(dir_relative_path, image_file_name))
+    if os.path.isfile(os.path.join(dir_path, reduced_file_name)):
+        set_func('reduced_representation_relative_path',
+                 os.path.join(dir_relative_path, reduced_file_name))
+    if os.path.isfile(os.path.join(dir_path, words_file_name)):
+        set_func('neurosynth_words_relative_path',
+                 os.path.join(dir_relative_path, words_file_name))
+    loaded = _add_absolute_paths(
+        os.path.dirname(dir_path), loaded, force=force)
+    return loaded
+
+
+def _managed_method(attr='download_manager_'):
```

Comment:
Such sophistatication makes the code hard to read. Wouldn't it be possible to inline this decorator: take its logic and move it to the various places that use it (I have found 3 places, so the amount of duplication won't be large)